diff --git a/drivers/dma/Kconfig b/drivers/dma/Kconfig
index 90284ffda58a..54e6e748205a 100644
--- a/drivers/dma/Kconfig
+++ b/drivers/dma/Kconfig
@@ -248,6 +248,15 @@ config HISI_DMA
 	help
 	  Support HiSilicon Kunpeng DMA engine.
 
+config IIS_IDMA
+	tristate "IIS iDMA Engine support"
+	depends on RISCV || COMPILE_TEST
+	depends on OF
+	select DMA_ENGINE
+	select DMA_VIRTUAL_CHANNELS
+	help
+	  Enable support for the IIS intelligent DMA controller (iDMA).
+
 config IMG_MDC_DMA
 	tristate "IMG MDC support"
 	depends on MIPS || COMPILE_TEST
